### ISSUE TITLE
I can't choose another proxy

### ISSUE THREAD

**Issue by Psychopath575** [foxyproxy/browser-extension#76]
Hello. I can't choose another proxy. When choosing another proxy, only the topmost one is selected, the first one. If you swap, then the desired proxy is selected. How to solve the problem? The screenshot shows that I have chosen a different country, but Italy still works
![image](https://github.com/foxyproxy/browser-extension/assets/153503814/fcb2bd52-4266-4ee2-a655-bedfa66fb7e4)

**Comment by Psychopath575** [comment]
![image](https://github.com/foxyproxy/browser-extension/assets/153503814/9ff89e8d-2235-40ac-8bf6-b6eee929745a)

**Comment by erosman** [comment]
Sometimes, browser caches the extension files and they are not properly updated during an extension upgrade.

## Step 1

- Disable extension
- Restart browser
- Enable extension

## Step 2
If step 1 didn't help...

- Export FoxyProxy settings
- Remove FoxyProxy
- Restart browser to clear its cache
- Reinstall FoxyProxy
- Import your settings

**Comment by Psychopath575** [comment]
> Sometimes, browser caches the extension files and they are not properly updated during an extension upgrade.
> 
> ## Step 1
> * Disable extension
> * Restart browser
> * Enable extension
> 
> ## Step 2
> If step 1 didn't help...
> 
> * Export FoxyProxy settings
> * Remove FoxyProxy
> * Restart browser to clear its cache
> * Reinstall FoxyProxy
> * Import your settings

This solution did not help me

**Comment by erosman** [comment]
Please try the following ...

- Go to `about:debugging#/runtime/this-firefox`
- Find FoxyProxy
- Click **Inspect**
- Try to change the proxy from the toolbar popup
- See if there are any errors showing in the Extension Developer Tools

**Comment by erosman** [comment]
Do the proxies have the same hostname & port?

**Comment by Psychopath575** [comment]
> Do the proxies have the same hostname & port?

Yes!

**Comment by erosman** [comment]
That is the reason. 
When FoxyProxy looks for proxies, it checks their hostname & port to find the right one.

What is the reason for having multiple proxies for the same server?

**Comment by Psychopath575** [comment]
> That is the reason. When FoxyProxy looks for proxies, it checks their hostname & port to find the right one.
> 
> What is the reason for having multiple proxies for the same server?

I use resident proxies from iproyal.
Why did everything work fine before?

**Comment by Psychopath575** [comment]
And how to solve the problem?

**Comment by erosman** [comment]
> Why did everything work fine before?

FoxyProxy 8 was written from the ground up, so the code is not the same as v7.

- How does IPRoyal work in selecting the right proxy/location?
- Are you logging in to the same server with multiple accounts and login credentials?
- What type of proxies are they? (http/https/socks etc)

**Comment by Psychopath575** [comment]
> > Why did everything work fine before?
> 
> FoxyProxy 8 was written from the ground up, so the code is not the same as v7.
> 
> * How does IPRoyal work in selecting the right proxy/location?
> * Are you logging in to the same server with multiple accounts and login credentials?
> * What type of proxies are they? (http/https/socks etc)

The same data is used there, only the password differs

**Comment by erosman** [comment]
I would need to work on it.
In the meantime, you can downgrade to v7.5.1
https://github.com/foxyproxy/browser-extension#2-downgrade-to-7

**Comment by erosman** [comment]
> The same data is used there, only the password differs

So the username is the same but the password is different?
There is another person who had a case where the usernames where different.

**Comment by Psychopath575** [comment]
> > The same data is used there, only the password differs
> 
> So the username is the same but the password is different? There is another person who had a case where the usernames where different.

Yes. Only the password is different.

**Comment by erosman** [comment]
Repo is updated with v8.8 and a workaround for this issue.

Please read [Help: Proxies](https://foxyproxy.github.io/browser-extension/src/content/help.html#proxies) for details.

Let ms know if it works out.

**Comment by erosman** [comment]
Repo is updated with v8.8 and a workaround for this issue.

Please read [Help: Proxies](https://foxyproxy.github.io/browser-extension/src/content/help.html#proxies) for details.

Let me know how it works out.

**Comment by erosman** [comment]
v8.8 was uploaded to AMO.
FoxyProxy Basic is available now and FoxyProxy Standard will be available after the AMO review.

**Comment by yunyty** [comment]
> Repo is updated with v8.8 and a workaround for this issue.
> 
> Please read [Help: Proxies](https://foxyproxy.github.io/browser-extension/src/content/help.html#proxies) for details.
> 
> Let me know how it works out.

I tried v8.8, but the issue is still there
For your reference: https://www.youtube.com/watch?v=kbHmPF4oqEY

Proxy List:
```
geo.iproyal.com:32325:UsernameABC:PasswordABC_country-tr
geo.iproyal.com:32325a:UsernameABC:PasswordABC_country-in
geo.iproyal.com:32325b:UsernameABC:PasswordABC_country-pk
geo.iproyal.com:32325d:UsernameABC:PasswordABC_country-gb

**Comment by erosman** [comment]
I see... The updated process fixed when in "Proxy by Patterns" is selected.
There is a conflict when a single proxy is selected with a fix for #47
I will fix that.

**Comment by erosman** [comment]
I have updated the code. It should be fixed for v8.9.

**Comment by yunyty** [comment]
> I have updated the code. It should be fixed for v8.9.

Can I get v8.9?

**Comment by erosman** [comment]
> Can I get v8.9?

Do you want to test the 8.9 beta?

Here is the latest beta.

[foxyproxy.8.9.beta.20240108.zip](https://github.com/foxyproxy/browser-extension/files/13862746/foxyproxy.8.9.beta.20240108.zip)

- Export your settings before installing
- Unsigned addons can only be installed in the Firefox Nightly or the Developer Edition

**Comment by yunyty** [comment]
Hello @erosman 

I tested v8.9 and found that the issue with Iproyal Sock5 has been resolved. However, there still appears to be an issue with the HTTP type. Please refer to the video I have provided for more information: https://www.youtube.com/watch?v=e-LaQfTAO3k

Import Proxy List:
```
Type: HTTP
rp.proxyscrape.com:6060a:UsernameABC-country-tr:PasswordABC
rp.proxyscrape.com:6060b:UsernameABC-country-in:PasswordABC
rp.proxyscrape.com:6060c:UsernameABC-country-pk:PasswordABC

**Comment by erosman** [comment]
@yunyty Thank you for testing the issue with v8.9.

AFA the video, can you close the `https://getfoxyproxy.org/geoip/` and reopen it in a new tab?
Alternatively, can you test using the "IP" button (or any other location testing site)?
Let me know the result.

**Comment by yunyty** [comment]
> @yunyty Thank you for testing the issue with v8.9.
> 
> AFA the video, can you close the `https://getfoxyproxy.org/geoip/` and reopen it in a new tab? Alternatively, can you test using the "IP" button (or any other location testing site)? Let me know the result.

Thank you for your attention to this matter. Here are the results: https://www.youtube.com/watch?v=KR8o_d6FN1Q

**Comment by erosman** [comment]
I see... there is a problem somewhere. I will look into it.
Can you also check the log while you change proxies?
It should show that the proxy is changed.
You can test it which any site e.g. `example.com` and check log result.
The Log shows what browser is actually doing.

**Comment by yunyty** [comment]
> I see... there is a problem somewhere. I will look into it. Can you also check the log while you change proxies? It should show that the proxy is changed. You can test it which any site e.g. `example.com` and check log result. The Log shows what browser is actually doing.

I opened the log tab and visited whatismyip.com without any active proxies. The Type/Hostname/Port was empty. 

Next, I changed the proxies and refreshed the page. This time, the Type/Hostname/Port was HTTP/rp.proxyscrape.com/6060. 

After that, I changed to another proxy and refreshed the page again. However, the Type/Hostname/Port was still the same as before - HTTP/rp.proxyscrape.com/6060. 

Would you like me to display the log for you?

**Comment by erosman** [comment]
> Would you like me to display the log for you?

No, thank you.
I will re-check the code.

**Comment by erosman** [comment]
I found out the issue. It is the authentication i.e. sending user/pass.

For SOCKS & HTTPS, FoxyProxy sends the authentication along with the request.
Unfortunately, the API does not allow sending user/pass for HTTP.  
I have already requested it in 2022 in [Bug 1794464: Allow HTTP authentication in proxy.onRequest](https://bugzilla.mozilla.org/show_bug.cgi?id=1794464).

For HTTP, the authentication depends on [407 Proxy Authentication Required](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/407).
The 407 Proxy Authentication request comes in form of `host` & `port`.
In this case, there is no way to know which user/pass to send, therefore the authentication for the first matching one is sent.

I cant think of a way around this at the moment, since that is how "407 Proxy Authentication" works.

> [!Note]
> For now, multiple proxies with the same host & port will only work for SOCKS & HTTPS.

I will update the Help accordingly.
